x-pack/plugins/fleet/server/services/epm/package_service.ts

```diff
@@ -263,6 +275,27 @@ class PackageClientImpl implements PackageClient {
     return generatePackageInfoFromArchiveBuffer(archiveBuffer, 'application/zip');
   }
 
+  public async getAgentPolicyInputs(pkgName: string, pkgVersion?: string, prerelease?: false) {
+    await this.#runPreflight(READ_PACKAGE_INFO_AUTHZ);
+
+    // If pkgVersion isn't specified, find the latest package version
+    if (!pkgVersion) {
+      const pkg = await Registry.fetchFindLatestPackageOrThrow(pkgName, { prerelease });
+      pkgVersion = pkg.version;
+    }
+
+    const { inputs } = await getTemplateInputs(
+      this.internalSoClient,
+      pkgName,
+      pkgVersion,
+      'json',
+      prerelease
+    );
+
+    // The shape of inputs returned by `getTemplateInputs` does not seem to match its type definition so need to manually cast it
+    return inputs as unknown as Array<Partial<FullAgentPolicyInput>>;
```

Comment:
I think there might be a typing issue here with `getTemplateInputs` returning an incorrect type. 

Here is an example input returned by `getTemplateInputs`:

```json
{
    "id": "logfile-apache",
    "type": "logfile",
    "streams": [
      {
        "id": "logfile-apache.access",
        "data_stream": {
          "type": "logs",
          "dataset": "apache.access"
        },
        "paths": [
          "/var/log/apache2/access.log*",
          "/var/log/apache2/other_vhosts_access.log*",
          "/var/log/httpd/access_log*"
        ]
      }
    ]
  }
```

And here is the type definition of its return type:

```ts
export interface FullAgentPolicyInputStream {
  id: string;
  data_stream: {
    dataset: string;
    type: string;
  };
  [key: string]: any;
}
```

They clearly don't match. 

I think the real return type should be `FullAgentPolicyInput`, not `FullAgentPolicyInputStream` but I'm also not sure since there's no complete reference for the agent settings (`inputs[].streams` isn't documented anywhere) so I'm finding it difficult to work out what settings exist at which level and what the intend behind the function is. 

I believe it is returning a top level input but it could also be returning an input stream. Either way its current type does not match the shape of the object it returns for me so I am casting it to a top level `FullAgentPolicyInput`.

Reply:
> I  think the real return type should be FullAgentPolicyInput, not FullAgentPolicyInputStream

This looks like a valid assumption to me. 

> I'm also not sure since there's no complete reference for the agent settings (inputs[].streams appears not to be documented anywhere

This method didn't exist until https://github.com/elastic/kibana/pull/168015, so there's some context on that PR. 

What I think happened here is we "fixed" `getTemplateInputs` to return a top-level `input` instead of just the `streams` entries to address https://github.com/elastic/kibana/issues/180307 but did not add explicit type assertions and thus we have this disconnect. The original implementation returned an input stream which was unexpected - it should return a top-level input and should be typed as such. The fact that it doesn't looks like we're falling back to TS inference improperly here.

Reply:
@kpollich Is this the correct method to use for our use case at all? I'm slightly confused about the term template. In what way do consumers need to parse or evaluate this template further to get the final Agent config? 

In the onboarding flow we will detect which integrations (e.g. nginx, apache) a user has installed on their host by matching against a list of known file paths. We then install the Agent on the host system and configure it to ingest these logs. What we'd like is to generate the actual config for the user which will configure the Agent correctly to ingest these logs. This should be the real config, not a template where the user has to follow additional steps.

Reply:
Kyle can give more color to this, but by my understanding this is the right method - it will give us the "default" config which is what we want here as the goal is to not require the user to configure anything.

They would need to change the template further to make it fit their needs (e.g. adding credentials for metrics etc.)

Reply:
Thanks for explanation @kpollich that's really helpful. 
Do you have a reference for these user variables? 
I am wondering if we need to add some logic to automatically evaluate them for the user during the onboarding flow.

Reply:
> I think the real return type should be `FullAgentPolicyInput` 

@thomheymann I think you're right, this type should be based off `FullAgentPolicyInput`: [here](https://github.com/elastic/kibana/blob/3ec981d03684634827779a882946b3fff90ddef2/x-pack/plugins/fleet/server/services/epm/packages/get_template_inputs.ts#L38-L43) is where the response is built and is based off [this other function](https://github.com/elastic/kibana/blob/3ec981d03684634827779a882946b3fff90ddef2/x-pack/plugins/fleet/server/services/agent_policies/package_policies_to_agent_inputs.ts#L55-L67), which instead returns a `FullAgentPolicyInput`. So I think that the return type should be something like this:
``` ts
 export type TemplateAgentPolicyInput = Pick<FullAgentPolicyInput, 'id' | 'type' | 'streams'>;
```

> What I think happened here is we "fixed" getTemplateInputs to return a top-level input instead of just the streams entries to address #180307 but did not add explicit type assertions and thus we have this disconnect. The original implementation returned an input stream which was unexpected - it should return a top-level input and should be typed as such. The fact that it doesn't looks like we're falling back to TS inference improperly here.

That's exactly what happened. I too had hard time understanding the exact shape of those streams, they're not really documented anywhere so I struggled a bit with them. 

> my understanding is that these are considered "templates" within Fleet, because Fleet typically compiles a full agent policy from these inputs by injecting user-defined variables, etc. The "template" designation seeks to clarify in the Fleet code paths that these are "uncompiled" inputs as opposed to FullAgentPolicyInput which will not only have user variables on a given package policy evaluated, but various pieces of Fleet-generated metadata as well.

@kpollich that's correct, in fact this logic is based off [storedPackagePolicyToAgentInputs](https://github.com/elastic/kibana/blob/3ec981d03684634827779a882946b3fff90ddef2/x-pack/plugins/fleet/server/services/agent_policies/package_policies_to_agent_inputs.ts#L27), which is used to compile the `template` coming from an integration by adding variable names, etc. The compiled input is then used as a part of the full agent policy, i.e. the yaml that fleet ships to the agent.

Reply:
> Do you have a reference for these user variables?

The user variables in question are unique per integration. https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/dev_docs/integrations_overview.md describes some of the overall policy construction workflow as it relates to integrations and may be helpful here.

Reply:
> I am wondering if we need to add some logic to automatically evaluate them for the user during the onboarding flow to ensure that the Agent is correctly configured.

My understanding is that the default "evaluation" is that all the default values are used, which seems like what we want in this situation. At least that's what I'm seeing in the "configs" tab on the integrations page:

<img width="501" alt="Screenshot 2024-05-31 at 15 43 51" src="https://github.com/elastic/kibana/assets/1508364/7a4acf0a-ad06-49ba-8fac-b29f25ba9808">

Reply:
> My understanding is that the default "evaluation" is that all the default values are used, which seems like what we want in this situation. At least that's what I'm seeing in the "configs" tab on the integrations page:

This is the correct understanding. Default values defined by the integration's various `manifest.yml` files will be included in the return value of this method.

Reply:
Brilliant, thanks all, I'll update the type as per @criamico suggestion in that case.